In [1]:
from IPython.display import display, HTML
display(HTML("<style>.container { width:90% !important; }</style>"))

<a id='31-bullet'></a>

<b> Название проекта: </b>
    
Анализ базы данных   

<b> Данное исследование разделим на несколько частей:<b/>

##### Часть 1. Изучение общей информации:
* [1. Изучение файлов с данными, получение общей информации, загрузка библиотек.](#1-bullet)

##### Часть 2. EDA:
* [1. Посчитайте, сколько книг вышло после 1 января 2000 года;](#2.1-bullet)
* [2. Для каждой книги посчитайте количество обзоров и среднюю оценку;](#2.2-bullet)
* [3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;](#2.3-bullet)
* [4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;](#2.4-bullet)
* [5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.](#2.5-bullet)

##### Часть 6. Общий вывод:
* [1. Общий вывод и рекомендации.](#6-bullet)  
   

  
<b> Описание данных проекта:</b>
    
**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<a id='1-bullet'></a>
# Изучение общей информации:

## Изучение файлов с данными, получение общей информации, загрузка библиотек.

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config =  # имя пользователя
'pwd':  # пароль
'host': 
'port':  # порт подключения
'db':  # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
def sql(query):
    query = pd.io.sql.read_sql(query, con = engine)
    return query

In [4]:
books = """
SELECT * FROM books
"""
sql(books).info()
sql(books).head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
authors = """
SELECT * FROM authors
"""
sql(authors).info()
sql(authors).head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
publishers = """
SELECT * FROM publishers
"""
sql(publishers).info()
sql(publishers).head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
ratings = """
SELECT * FROM ratings
"""
sql(ratings).info()
sql(ratings).head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
reviews = """
SELECT * FROM reviews
"""
sql(reviews).info()
sql(reviews).head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [47]:
title_rating_review = """
        WITH
        rev AS(
                SELECT book_id, COUNT(review_id) AS review_count
                FROM reviews
                GROUP BY book_id),
                
        rat AS( SELECT book_id, AVG(rating) AS avg_rating
                FROM ratings
                GROUP BY book_id)
                
        SELECT title, rev.review_count, rat.avg_rating
        FROM books AS b
        LEFT JOIN rat USING(book_id)
        LEFT JOIN rev USING(book_id)        
        WHERE rev.review_count IS NOT NULL
        ORDER BY rev.review_count DESC
                """
sql(title_rating_review).head()

title  review_count  avg_rating
0                 Twilight (Twilight  #1)             7    3.662500
1                          The Book Thief             6    4.264151
2               Outlander (Outlander  #1)             6    4.125000
3                     Water for Elephants             6    3.977273
4  The Da Vinci Code (Robert Langdon  #2)             6    3.830508

<b> Вывод:
    
    Имеем 5 таблиц с информацией о книгах (books), авторов (authors), издательствах (publishers), рейтинге (ratings) и обзоров (reviews). Данные полные, но есть проблема с типом данных даты 'publication_date' в books, при работе нужно учитывать.

<a id='2-bullet'></a>
# Часть 2. EDA:

<a id='2.1-bullet'></a>
## Посчитайте, сколько книг вышло после 1 января 2000 года;

In [9]:
books_after_01_01_2000 = """
                        SELECT count(DISTINCT book_id) AS books_after_01_01_2000
                        FROM books
                        WHERE CAST( publication_date AS date) > '2000.01.01'
                        """
sql(books_after_01_01_2000)

books_after_01_01_2000
0                     819

<b> Вывод:</b>
819 книг вышло после 1 января 2000 года

<a id='2.2-bullet'></a>
## Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [53]:
title_rating_review = """
        WITH
        rev AS(
                SELECT book_id, COUNT(review_id) AS review_count
                FROM reviews
                GROUP BY book_id),
                
        rat AS( SELECT book_id, AVG(rating) AS avg_rating
                FROM ratings
                GROUP BY book_id)
                
        SELECT title, rev.review_count, rat.avg_rating
        FROM books AS b
        LEFT JOIN rat USING(book_id)
        LEFT JOIN rev USING(book_id)        
        WHERE review_count IS NOT NULL
        ORDER BY review_count DESC
                """
sql(title_rating_review).head()

title  review_count  avg_rating
0                 Twilight (Twilight  #1)             7    3.662500
1                          The Book Thief             6    4.264151
2               Outlander (Outlander  #1)             6    4.125000
3                     Water for Elephants             6    3.977273
4  The Da Vinci Code (Robert Langdon  #2)             6    3.830508

In [63]:
title_rating = """
        WITH
        rev AS(
                SELECT book_id, COUNT(review_id) AS review_count
                FROM reviews
                GROUP BY book_id),
                
        rat AS( SELECT book_id, AVG(rating) AS avg_rating
                FROM ratings
                GROUP BY book_id)
                
        SELECT title, rev.review_count, rat.avg_rating
        FROM books AS b
        LEFT JOIN rat USING(book_id)
        LEFT JOIN rev USING(book_id)        
        WHERE review_count IS NOT NULL
        ORDER BY avg_rating DESC
                """
sql(title_rating).head(43)

title  review_count  \
0           A Woman of Substance (Emma Harte Saga #1)             2   
1                Pop Goes the Weasel (Alex Cross  #5)             2   
2   In the Hand of the Goddess (Song of the Liones...             2   
3                Piercing the Darkness (Darkness  #2)             2   
4                     Act of Treason (Mitch Rapp  #9)             2   
5                  The Big Bad Wolf  (Alex Cross  #9)             2   
6                                My Name Is Asher Lev             2   
7   Wherever You Go  There You Are: Mindfulness Me...             2   
8                                          Dead Souls             2   
9                                               March             2   
10               Welcome to Temptation (Dempseys  #1)             2   
11                                              Women             2   
12            School's Out—Forever (Maximum Ride  #2)             3   
13                                      Alas  Babylon             2   
14  Crucial Conversations: Tools for Talking When ...             2   
15                          Triptych (Will Trent  #1)             2   
16               Stolen (Women of the Otherworld  #2)             2   
17                                          Geek Love             2   
18  How to Be a Domestic Goddess: Baking and the A...             1   
19  Captivating: Unveiling the Mystery of a Woman'...             2   
20  The Adventures of Tom Sawyer and Adventures of...             1   
21                                      Evening Class             2   
22                                 The Woman in Black             2   
23  The Ghost Map: The Story of London's Most Terr...             2   
24  Stone of Farewell (Memory  Sorrow  and Thorn  #2)             1   
25                                    Light in August             2   
26                                         Hard Times             2   
27                                        Angels Fall             2   
28       Moneyball: The Art of Winning an Unfair Game             3   
29                      Emily of New Moon (Emily  #1)             2   
30              A Fistful of Charms (The Hollows  #4)             2   
31      Arrows of the Queen (Heralds of Valdemar  #1)             2   
32                           Neil Gaiman's Neverwhere             2   
33                       Plum Island (John Corey  #1)             2   
34  The Walking Dead  Book One (The Walking Dead #...             2   
35  The Demon-Haunted World: Science as a Candle i...             2   
36                  Misty of Chincoteague (Misty  #1)             2   
37  The War of Art: Break Through the Blocks & Win...             2   
38                      A Dirty Job (Grim Reaper  #1)             4   
39           Evil Under the Sun (Hercule Poirot  #24)             2   
40                                        Marvel 1602             2   
41                                Homage to Catalonia             2   
42                           Tai-Pan (Asian Saga  #2)             2   

    avg_rating  
0          5.0  
1          5.0  
2          5.0  
3          5.0  
4          5.0  
5          5.0  
6          5.0  
7          5.0  
8          5.0  
9          5.0  
10         5.0  
11         5.0  
12         5.0  
13         5.0  
14         5.0  
15         5.0  
16         5.0  
17         5.0  
18         5.0  
19         5.0  
20         5.0  
21         5.0  
22         5.0  
23         5.0  
24         5.0  
25         5.0  
26         5.0  
27         5.0  
28         5.0  
29         5.0  
30         5.0  
31         5.0  
32         5.0  
33         5.0  
34         5.0  
35         5.0  
36         5.0  
37         5.0  
38         5.0  
39         5.0  
40         5.0  
41         5.0  
42         5.0

In [52]:
title_no_review = """
        WITH
        rev AS(
                SELECT book_id, COUNT(review_id) AS review_count
                FROM reviews
                GROUP BY book_id),
                
        rat AS( SELECT book_id, AVG(rating) AS avg_rating
                FROM ratings
                GROUP BY book_id)
                
        SELECT title, rev.review_count, rat.avg_rating
        FROM books AS b
        LEFT JOIN rat USING(book_id)
        LEFT JOIN rev USING(book_id)
        WHERE review_count IS NULL
        ORDER BY avg_rating DESC
                """
sql(title_no_review)

title review_count  avg_rating
0   The Cat in the Hat and Other Dr. Seuss Favorites         None    5.000000
1                               Leonardo's Notebooks         None    4.000000
2  Disney's Beauty and the Beast (A Little Golden...         None    4.000000
3                          Essential Tales and Poems         None    4.000000
4    Anne Rice's The Vampire Lestat: A Graphic Novel         None    3.666667
5                            The Natural Way to Draw         None    3.000000

<b> Вывод:</b>
В таблице для каждой книги посчитано количество обзоров и средняя оценка. 
6  книг не имеют обзоров:
* The Cat in the Hat and Other Dr. Seuss Favorites
* Leonardo's Notebooks	
* Disney's Beauty and the Beast
* Essential Tales and Poems	
* Anne Rice's The Vampire Lestat: A Graphic Novel
* The Natural Way to Draw
По количеству обзоров лидирует: Twilight (Twilight #1)
По рейтингу 43 книги имеют среднюю оценку 5

<a id='2.3-bullet'></a>
## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [101]:
top_publisher = """
                SELECT publisher, COUNT(title) as book_count
                FROM publishers as p
                LEFT JOIN  books AS b ON p.publisher_id = b.publisher_id
                WHERE num_pages > 50
                GROUP BY publisher
                ORDER BY book_count DESC
                LIMIT 1
                """
sql(top_publisher)

publisher  book_count
0  Penguin Books          42

<b> Вывод:</b>
издательство, которое выпустило наибольшее число книг толще 50 страниц: Penguin Books

<a id='2.4-bullet'></a>
## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [12]:
top_author = """
        WITH
        title AS(
                SELECT title, b.author_id, count(rating) AS count_rating
                FROM books as b
                JOIN ratings AS rat ON b.book_id = rat.book_id
                GROUP BY title, b.author_id
                HAVING count(rating) > 50
                ORDER BY count_rating DESC),
        autor AS(
                SELECT author, b.author_id, AVG(rating) AS avg_rating
                FROM books as b
                JOIN authors AS a ON b.author_id = a.author_id
                JOIN ratings AS rat ON b.book_id = rat.book_id
                GROUP BY author, b.author_id
                ORDER BY avg_rating DESC)
                
        SELECT author, avg_rating
        FROM title
        JOIN autor USING(author_id)
        ORDER BY avg_rating DESC
        LIMIT 1
                """
sql(top_author)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.288462

<b> Вывод:</b>
Автор с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками: J.K. Rowling/Mary GrandPré	

<a id='2.5-bullet'></a>
## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [13]:
avg_reviews = """
        WITH
        rev AS (SELECT username, COUNT(text) AS review_number
                FROM reviews
                GROUP BY username
                ORDER BY review_number DESC),
                
        rat AS (SELECT username, COUNT(rating) AS rating_number
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating) > 50
                ORDER BY rating_number DESC)
                
        SELECT AVG(rev.review_number) AS avg_review_number
        FROM rev 
        JOIN rat USING(username)
                """
sql(avg_reviews)

avg_review_number
0          24.333333

<b> Вывод:</b>
Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24

<a id='6-bullet'></a>
# Часть 3. Общий вывод:

<b> Из предложенных данных, мы обнаружили: </b>
* Имеем 5 таблиц с информацией о книгах (books), авторов (authors), издательствах (publishers), рейтинге (ratings) и обзоров (reviews). 
* Данные полные, но есть проблема с типом данных даты 'publication_date' в books. При работе учлел.
* 819 книг вышло после 1 января 2000 года.
* В таблице для каждой книги посчитано количество обзоров и средняя оценка. 
    * 6  книг не имеют обзоров:
        * The Cat in the Hat and Other Dr. Seuss Favorites
        * Leonardo's Notebooks	
        * Disney's Beauty and the Beast
        * Essential Tales and Poems	
        * Anne Rice's The Vampire Lestat: A Graphic Novel
        * The Natural Way to Draw
    * По количеству обзоров лидирует: Twilight (Twilight #1)
    * По рейтингу 43 книги имеют среднюю оценку 5
* Издательство, которое выпустило наибольшее число книг толще 50 страниц: Penguin Books
* Автор с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками: J.K. Rowling/Mary GrandPré
* Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24

# [Вернуться к началу.](#31-bullet)